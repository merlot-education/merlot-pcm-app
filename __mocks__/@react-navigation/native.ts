const useFocusEffect = jest.fn()

export default { useFocusEffect }
