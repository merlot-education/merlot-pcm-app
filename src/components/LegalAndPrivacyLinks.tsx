import React from 'react';
import { Trans, useTranslation } from 'react-i18next';
import { StyleSheet, View, Text, Linking } from 'react-native';
import { ColorPallet, TextTheme } from '../theme/theme';

const LegalAndPrivacyLinks = () => {
  const { t } = useTranslation();
  const openLink = (url: string) => {
    Linking.canOpenURL(url).then(supported => {
      if (supported) {
        Linking.openURL(url);
      } else {
        console.log("Don't know how to open URI: " + url);
      }
    });
  };
  return (
    <View>
      <View style={styles.acceptTerms}>
        <Text style={styles.termsText}>
          <Trans
            i18nKey="Terms.AcceptTerms"
            components={{
              apachelink: (
                <Text
                  onPress={() => openLink('https://gitlab.eclipse.org/eclipse/xfsc/pcm/app/-/blob/main/LICENSE')}
                  style={styles.linkText}
                />
              ),
              privacylink: (
                <Text
                  onPress={() => openLink('https://gitlab.eclipse.org/eclipse/xfsc/pcm/app/-/blob/main/GDPR.md')}
                  style={styles.linkText}
                />
              ),
            }}
          />
        </Text>
      </View>
      <View style={styles.acceptTermsContinue}>
        <Text style={styles.termsText}>
          {t<string>('Terms.AcceptTermsContinue')}
        </Text>
      </View>
    </View>
  );
};

export default LegalAndPrivacyLinks;

const styles = StyleSheet.create({
  acceptTerms: {
    paddingHorizontal: 20,
    paddingTop: 20,
  },
  acceptTermsContinue: {
    paddingTop: 8,
    paddingHorizontal: 20,
    paddingBottom: 20,
  },
  termsText: {
    ...TextTheme.normal,
    fontWeight: 'bold',
    textAlign: 'justify',
  },
  link: {
    marginVertical: 14,
    padding: 20,
    alignItems: 'center',
  },
  linkText: {
    ...TextTheme.normal,
    fontWeight: 'bold',
    color: ColorPallet.brand.primary,
    textDecorationLine: 'underline',
  },
});
