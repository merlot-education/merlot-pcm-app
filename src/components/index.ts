import HighlightTextBox from './text/HighlightTextBox';
import Label from './text/Label';
import Text from './text/Text';
import Title from './text/Title';
import TextInput from './inputs/TextInput';
import ContactListItem from './listItems/ContactListItem';
import NotificationListItem from './listItems/NotificationListItem';
import SettingListItem from './listItems/SettingListItem';
import Loader from './loader/Loader';
import SafeAreaScrollView from './views/SafeAreaScrollView';
import SingleSelectBlock from './inputs/SingleSelectBlock';
import IconButton from './iconButton/IconButton';
import ScreenNavigatorButtons from './screenNavigatorButtons/ScreenNavigatorButtons';
import InfoCard from './cards/InfoCard';

export {
  HighlightTextBox,
  Label,
  Text,
  Title,
  TextInput,
  ContactListItem,
  NotificationListItem,
  Loader,
  SettingListItem,
  SafeAreaScrollView,
  SingleSelectBlock,
  IconButton,
  ScreenNavigatorButtons,
  InfoCard,
};
