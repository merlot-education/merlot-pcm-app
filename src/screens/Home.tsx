import React from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { useTranslation } from 'react-i18next'
import { Colors, TextTheme } from '../theme/theme'
import Button, { ButtonType } from '../components/button/Button'
import Screens from '../utils/constants'

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.background,
    margin: 20,
  },
  bodyText: {
    ...TextTheme.normal,
    flexShrink: 1,
  },
  spacer: {
    height: 40,
  },
})

interface HomeProps {
  navigation: any
}

const Home: React.FC<HomeProps> = ({ navigation }) => {
  const { t } = useTranslation()

  const nextPage = () => {
    navigation.navigate(Screens.Scan)
  }

  return (
    <View style={[styles.container]}>
      <Text style={[styles.bodyText, { fontWeight: 'bold' }]}>
        {/* {t('Home.LoginMsg')} */}
        Your wallet is created successfully.
      </Text>
      <View style={styles.spacer} />
      <Button
        title={t('Global.Continue')}
        buttonType={ButtonType.Primary}
        onPress={nextPage}
      />
    </View>
  )
}

export default Home
