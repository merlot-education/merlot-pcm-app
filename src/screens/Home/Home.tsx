import React from 'react';
import {
  Dimensions,
  FlatList,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from 'react-native';
import { useTranslation } from 'react-i18next';
import { StackScreenProps } from '@react-navigation/stack';
import { CredentialState } from '@aries-framework/core';
import { useCredentialByState } from '@aries-framework/react-hooks';
import { ColorPallet, TextTheme } from '../../theme/theme';
import useNotifications from '../../hooks/notifications';
import { HomeStackParams, Screens } from '../../types/navigators';
import InfoTextBox from '../../components/text/InfoTextBox';
import { NotificationListItem } from '../../components';
import { NotificationType } from '../../components/listItems/NotificationListItem';

const { width } = Dimensions.get('window');

type HomeProps = StackScreenProps<HomeStackParams, Screens.Home>;

const Home: React.FC<HomeProps> = ({ navigation }) => {
  const { t } = useTranslation();
  const { notifications } = useNotifications();
  const credentials = [
    ...useCredentialByState(CredentialState.CredentialReceived),
    ...useCredentialByState(CredentialState.Done),
  ];
  const emptyListComponent = () => (
    <View style={{ marginHorizontal: offset, width: width - 2 * offset }}>
      <InfoTextBox showIcon>
        <Text style={TextTheme.normal}>{t('Home.NoNewUpdates')}</Text>
      </InfoTextBox>
    </View>
  );

  const displayMessage = (credentialCount: number) => {
    if (typeof credentialCount === 'undefined') {
      throw new Error(t('Home.CredentialCountUndefinedError'));
    }

    let credentialMsg;

    if (credentialCount === 1) {
      credentialMsg = (
        <Text>
          {t('Home.YouHave')}{' '}
          <Text style={{ fontWeight: 'bold' }}>{credentialCount}</Text>{' '}
          {t('Home.Credential')} {t('Home.InYourWallet')}
        </Text>
      );
    } else if (credentialCount > 1) {
      credentialMsg = (
        <Text>
          {t('Home.YouHave')}{' '}
          <Text style={{ fontWeight: 'bold' }}>{credentialCount}</Text>{' '}
          {t('Home.Credentials')} {t('Home.InYourWallet')}
        </Text>
      );
    } else {
      credentialMsg = t('Home.NoCredentials');
    }

    return (
      <View style={[styles.messageContainer]}>
        {credentialCount === 0 ? (
          <Text
            style={[
              TextTheme.headingTwo,
              { marginTop: offset, marginBottom: 35 },
            ]}>
            {t('Home.Welcome')}
          </Text>
        ) : null}
        <Text
          style={[
            TextTheme.normal,
            { marginTop: offset, textAlign: 'center' },
          ]}>
          {credentialMsg}
        </Text>
      </View>
    );
  };

  return (
    <View>
      <View style={styles.rowContainer}>
        <Text style={[TextTheme.headingFour, styles.header]}>
          {t('Home.Notifications')}
          {notifications.length ? ` (${notifications.length})` : ''}
        </Text>
        {notifications?.length > 1 ? (
          <TouchableOpacity
            style={styles.linkContainer}
            activeOpacity={1}
            onPress={() => navigation.navigate(Screens.Notifications)}>
            <Text style={styles.link}>{t('Home.SeeAll')}</Text>
          </TouchableOpacity>
        ) : null}
      </View>
      <FlatList
        horizontal
        showsHorizontalScrollIndicator={false}
        scrollEnabled={notifications.length > 0}
        snapToOffsets={[
          0,
          ...Array(notifications.length)
            .fill(0)
            .map(
              (n: number, i: number) =>
                i * (width - 2 * (offset - offsetPadding)),
            )
            .slice(1),
        ]}
        decelerationRate="fast"
        ListEmptyComponent={emptyListComponent()}
        data={notifications}
        keyExtractor={item => item.id}
        renderItem={({ item, index }) => (
          <View
            style={{
              width: width - 3.5 * offset,
              marginLeft: !index ? offset : offsetPadding,
              marginRight:
                index === notifications.length - 1 ? offset : offsetPadding,
            }}>
            {item.type === 'CredentialRecord' ? (
              <NotificationListItem
                key={item.id}
                notificationType={NotificationType.CredentialOffer}
                notification={item}
              />
            ) : (
              <NotificationListItem
                key={item.id}
                notificationType={NotificationType.ProofRequest}
                notification={item}
              />
            )}
          </View>
        )}
      />
      <View style={styles.container}>{displayMessage(credentials.length)}</View>
    </View>
  );
};

export default Home;

const offset = 25;
const offsetPadding = 5;

const styles = StyleSheet.create({
  container: {
    paddingHorizontal: offset,
  },
  rowContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: offset,
  },
  messageContainer: {
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 35,
    marginHorizontal: 60,
  },
  header: {
    marginTop: offset,
    marginBottom: 20,
  },
  linkContainer: {
    minHeight: TextTheme.normal.fontSize,
    marginTop: 10,
  },
  link: {
    ...TextTheme.normal,
    color: ColorPallet.brand.link,
  },
});
