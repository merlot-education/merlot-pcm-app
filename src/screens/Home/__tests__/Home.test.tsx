import { useNavigation } from '@react-navigation/core'
import { render, waitFor } from '@testing-library/react-native'
import React from 'react'
import { FlatList } from 'react-native'
import { create } from 'react-test-renderer'
import { NotificationListItem } from '../../../components'
import Home from '../Home'

describe('displays a home screen', () => {
  jest.useFakeTimers()
  it('renders correctly', () => {
    const tree = create(<Home navigation={useNavigation()} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  /**
   * Scenario: Home Screen without any pending notification
   * Given wallet has successfully loaded
   * When the holder selects the "Home" button in the main navigation bar
   * Then the Home Screen is displayed
   */
  it('defaults to no notifications', async () => {
    const { findByText } = render(<Home navigation={useNavigation()} />)
    const notificationLabel = await findByText('Home.NoNewUpdates')

    expect(notificationLabel).toBeTruthy()
  })
})

describe('with a notifications module, when there are no notifications', () => {
  jest.useFakeTimers()
  it('notifications are empty', async () => {
    const tree = create(<Home navigation={useNavigation()} />)
    const { root } = tree
    const flatListInstance = root.findByType(FlatList)

    await waitFor(() =>
      expect(flatListInstance.findAllByType(NotificationListItem)).toHaveLength(
        0,
      ),
    )
  })
})
