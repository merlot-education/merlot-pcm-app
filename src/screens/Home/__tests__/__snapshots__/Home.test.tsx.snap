// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a home screen renders correctly 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 25,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "TitilliumWeb-SemiBold",
            "fontSize": 21,
            "fontWeight": "bold",
          },
          Object {
            "marginBottom": 20,
            "marginTop": 25,
          },
        ]
      }
    >
      Home.Notifications
      
    </Text>
  </View>
  <RCTScrollView
    ListEmptyComponent={
      <View
        style={
          Object {
            "marginHorizontal": 25,
            "width": 700,
          }
        }
      >
        <InfoTextBox
          showIcon={true}
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 17,
                "fontWeight": "normal",
              }
            }
          >
            Home.NoNewUpdates
          </Text>
        </InfoTextBox>
      </View>
    }
    data={Array []}
    decelerationRate="fast"
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    showsHorizontalScrollIndicator={false}
    snapToOffsets={
      Array [
        0,
      ]
    }
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "marginHorizontal": 25,
            "width": 700,
          }
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "#F2F2F2",
              "borderRadius": 5,
              "flexDirection": "row",
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignSelf": "center",
                "marginRight": 10,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/info-icon.png",
                }
              }
              style={
                Object {
                  "height": 30,
                  "width": 30,
                }
              }
            />
          </View>
          <Text
            accessibilityLabel="InfoTextBox"
            style={
              Object {
                "alignSelf": "center",
                "color": "#000000",
                "flexShrink": 1,
                "fontFamily": "TitilliumWeb-Regular",
                "fontSize": 17,
                "fontWeight": "normal",
              }
            }
            testID="InfoTextBox"
          >
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontFamily": "TitilliumWeb-Regular",
                  "fontSize": 17,
                  "fontWeight": "normal",
                }
              }
            >
              Home.NoNewUpdates
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "paddingHorizontal": 25,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginHorizontal": 60,
            "marginTop": 35,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "TitilliumWeb-SemiBold",
              "fontSize": 32,
              "fontWeight": "bold",
            },
            Object {
              "marginBottom": 35,
              "marginTop": 25,
            },
          ]
        }
      >
        Home.Welcome
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "TitilliumWeb-Regular",
              "fontSize": 17,
              "fontWeight": "normal",
            },
            Object {
              "marginTop": 25,
              "textAlign": "center",
            },
          ]
        }
      >
        Home.NoCredentials
      </Text>
    </View>
  </View>
</View>
`;
